### ISSUE TITLE
A2DP in CAVA

### ISSUE THREAD

**Issue by pasimko** [karlstav/cava#122]
I have a pair of bluetooth headphones, which I use to listen to music through A2DP. Cava cannot detect these. What changes can I make to my config file to fix this?

**Comment by karlstav** [comment]
are they listed under `pacmd list-sinks`?

**Comment by karlstav** [comment]
@pasimko could you give me the output of  `pacmd list-sinks`, when playing audio through the headphones?

**Comment by pasimko** [comment]
index: 26
	name: <bluez_sink.16_05_03_17_15_B8>
	driver: <module-bluez5-device.c>
	flags: HARDWARE DECIBEL_VOLUME LATENCY 
	state: SUSPENDED
	suspend cause: IDLE 
	priority: 9030
	volume: front-left: 16907 /  26% / -35.30 dB,   front-right: 16907 /  26% / -35.30 dB
	        balance 0.00
	base volume: 65536 / 100% / 0.00 dB
	volume steps: 65537
	muted: no
	current latency: 0.00 ms
	max request: 4 KiB
	max rewind: 0 KiB
	monitor source: 30
	sample spec: s16le 2ch 44100Hz
	channel map: front-left,front-right
	             Stereo
	used by: 0
	linked by: 0
	fixed latency: 48.22 ms
	card: 2 <bluez_card.16_05_03_17_15_B8>
	module: 21
	properties:
		bluetooth.protocol = "a2dp_sink"
		device.description = "Bluedio"
		device.string = "16:05:03:17:15:B8"
		device.api = "bluez"
		device.class = "sound"
		device.bus = "bluetooth"
		device.form_factor = "headset"
		bluez.path = "/org/bluez/hci0/dev_16_05_03_17_15_B8"
		bluez.class = "0x240404"
		bluez.alias = "Bluedio"
		device.icon_name = "audio-headset-bluetooth"
		device.intended_roles = "phone"
	ports:
		headset-output: Headset (priority 0, latency offset 0 usec, available: unknown)
			properties:
				
	active port: <headset-output>

**Comment by karlstav** [comment]
is that all? I thought that it would be more. Cava uses the default (active) output to find out what it should use as input. The current default output should have a `*` before the `index: n`. Can you give me the  output of `pacmd list-sources` as well?

**Comment by pasimko** [comment]
3 source(s) available.
    index: 0
	name: <alsa_output.pci-0000_00_1f.3.analog-stereo.monitor>
	driver: <module-alsa-card.c>
	flags: DECIBEL_VOLUME LATENCY DYNAMIC_LATENCY
	state: RUNNING
	suspend cause: 
	priority: 1950
	volume: front-left: 65536 / 100% / 0.00 dB,   front-right: 65536 / 100% / 0.00 dB
	        balance 0.00
	base volume: 65536 / 100% / 0.00 dB
	volume steps: 65537
	muted: no
	current latency: 0.00 ms
	max rewind: 64 KiB
	sample spec: s16le 2ch 44100Hz
	channel map: front-left,front-right
	             Stereo
	used by: 2
	linked by: 2
	configured latency: 40.00 ms; range is 0.50 .. 371.52 ms
	monitor_of: 0
	card: 0 <alsa_card.pci-0000_00_1f.3>
	module: 6
	properties:
		device.description = "Monitor of Built-in Audio Analog Stereo"
		device.class = "monitor"
		alsa.card = "0"
		alsa.card_name = "HDA Intel PCH"
		alsa.long_card_name = "HDA Intel PCH at 0xf7140000 irq 130"
		alsa.driver_name = "snd_hda_intel"
		device.bus_path = "pci-0000:00:1f.3"
		sysfs.path = "/devices/pci0000:00/0000:00:1f.3/sound/card0"
		device.bus = "pci"
		device.vendor.id = "8086"
		device.vendor.name = "Intel Corporation"
		device.product.id = "a170"
		device.product.name = "Sunrise Point-H HD Audio"
		device.form_factor = "internal"
		device.string = "0"
		module-udev-detect.discovered = "1"
		device.icon_name = "audio-card-pci"
  * index: 1
	name: <alsa_input.pci-0000_00_1f.3.analog-stereo>
	driver: <module-alsa-card.c>
	flags: HARDWARE HW_MUTE_CTRL HW_VOLUME_CTRL DECIBEL_VOLUME LATENCY DYNAMIC_LATENCY
	state: RUNNING
	suspend cause: 
	priority: 9959
	volume: front-left: 10093 /  15% / -48.75 dB,   front-right: 10093 /  15% / -48.75 dB
	        balance 0.00
	base volume: 6554 /  10% / -60.00 dB
	volume steps: 65537
	muted: no
	current latency: 5.04 ms
	max rewind: 0 KiB
	sample spec: s16le 2ch 44100Hz
	channel map: front-left,front-right
	             Stereo
	used by: 2
	linked by: 2
	configured latency: 40.00 ms; range is 0.50 .. 371.52 ms
	card: 0 <alsa_card.pci-0000_00_1f.3>
	module: 6
	properties:
		alsa.resolution_bits = "16"
		device.api = "alsa"
		device.class = "sound"
		alsa.class = "generic"
		alsa.subclass = "generic-mix"
		alsa.name = "ALC1150 Analog"
		alsa.id = "ALC1150 Analog"
		alsa.subdevice = "0"
		alsa.subdevice_name = "subdevice #0"
		alsa.device = "0"
		alsa.card = "0"
		alsa.card_name = "HDA Intel PCH"
		alsa.long_card_name = "HDA Intel PCH at 0xf7140000 irq 130"
		alsa.driver_name = "snd_hda_intel"
		device.bus_path = "pci-0000:00:1f.3"
		sysfs.path = "/devices/pci0000:00/0000:00:1f.3/sound/card0"
		device.bus = "pci"
		device.vendor.id = "8086"
		device.vendor.name = "Intel Corporation"
		device.product.id = "a170"
		device.product.name = "Sunrise Point-H HD Audio"
		device.form_factor = "internal"
		device.string = "front:0"
		device.buffering.buffer_size = "65536"
		device.buffering.fragment_size = "32768"
		device.access_mode = "mmap+timer"
		device.profile.name = "analog-stereo"
		device.profile.description = "Analog Stereo"
		device.description = "Built-in Audio Analog Stereo"
		alsa.mixer_name = "Realtek ALC1150"
		alsa.components = "HDA:10ec0900,104386ae,00100001 HDA:80862809,80860101,00100000"
		module-udev-detect.discovered = "1"
		device.icon_name = "audio-card-pci"
	ports:
		analog-input-front-mic: Front Microphone (priority 8500, latency offset 0 usec, available: no)
			properties:
				device.icon_name = "audio-input-microphone"
		analog-input-rear-mic: Rear Microphone (priority 8200, latency offset 0 usec, available: no)
			properties:
				device.icon_name = "audio-input-microphone"
		analog-input-linein: Line In (priority 8100, latency offset 0 usec, available: no)
			properties:
				
	active port: <analog-input-front-mic>
    index: 7
	name: <bluez_sink.16_05_03_17_15_B8.monitor>
	driver: <module-bluez5-device.c>
	flags: DECIBEL_VOLUME LATENCY 
	state: RUNNING
	suspend cause: 
	priority: 1030
	volume: front-left: 65536 / 100% / 0.00 dB,   front-right: 65536 / 100% / 0.00 dB
	        balance 0.00
	base volume: 65536 / 100% / 0.00 dB
	volume steps: 65537
	muted: no
	current latency: 0.00 ms
	max rewind: 0 KiB
	sample spec: s16le 2ch 44100Hz
	channel map: front-left,front-right
	             Stereo
	used by: 2
	linked by: 2
	fixed latency: 48.22 ms
	monitor_of: 4
	card: 3 <bluez_card.16_05_03_17_15_B8>
	module: 22
	properties:
		device.description = "Monitor of Bluedio"
		device.class = "monitor"
		device.string = "16:05:03:17:15:B8"
		device.api = "bluez"
		device.bus = "bluetooth"
		device.form_factor = "headset"
		bluez.path = "/org/bluez/hci0/dev_16_05_03_17_15_B8"
		bluez.class = "0x240404"
		bluez.alias = "Bluedio"
		device.icon_name = "audio-headset-bluetooth"
		device.intended_roles = "phone"

**Comment by karlstav** [comment]
I think I can see the poblem in you list of sinks:

```
* index: 0
name: <alsa_output.pci-0000_00_1f.3.analog-stereo>
```

the `*` indicates that this is selected as default. So cava will try to grab the audio being played to this output and not the Bluetooth headphone.

Try to go to your system sound settings and select the Bluetooth headphones in the output settings there. _Before_ launching cava.

**Comment by pasimko** [comment]
This fixed it. Thank you for your dedication!

**Comment by karlstav** [comment]
Great! Maybe I'll add something about this in the readme file.

**Comment by josswright** [comment]
I just came across this issue, and am also not getting any visualisation with my bluetooth headphones. I have run `pactl set-default-sink bluez_sink.04_52_C7_0A_32_96.a2dp_sink` before running cava.

`pacmd list-sinks`

`2 sink(s) available.
    index: 0
	name: <alsa_output.pci-0000_00_1f.3.analog-stereo>
	driver: <module-alsa-card.c>
	flags: HARDWARE HW_MUTE_CTRL HW_VOLUME_CTRL DECIBEL_VOLUME LATENCY FLAT_VOLUME DYNAMIC_LATENCY
	state: SUSPENDED
	suspend cause: IDLE
	priority: 9039
	volume: front-left: 26212 /  40% / -23.88 dB,   front-right: 26212 /  40% / -23.88 dB
	        balance 0.00
	base volume: 65536 / 100% / 0.00 dB
	volume steps: 65537
	muted: yes
	current latency: 0.00 ms
	max request: 0 KiB
	max rewind: 0 KiB
	monitor source: 0
	sample spec: s32le 2ch 44100Hz
	channel map: front-left,front-right
	             Stereo
	used by: 0
	linked by: 0
	configured latency: 0.00 ms; range is 0.50 .. 2000.00 ms
	card: 0 <alsa_card.pci-0000_00_1f.3>
	module: 6
	properties:
		alsa.resolution_bits = "32"
		device.api = "alsa"
		device.class = "sound"
		alsa.class = "generic"
		alsa.subclass = "generic-mix"
		alsa.name = "ALC298 Analog"
		alsa.id = "ALC298 Analog"
		alsa.subdevice = "0"
		alsa.subdevice_name = "subdevice #0"
		alsa.device = "0"
		alsa.card = "0"
		alsa.card_name = "HDA Intel PCH"
		alsa.long_card_name = "HDA Intel PCH at 0xe1340000 irq 136"
		alsa.driver_name = "snd_hda_intel"
		device.bus_path = "pci-0000:00:1f.3"
		sysfs.path = "/devices/pci0000:00/0000:00:1f.3/sound/card0"
		device.bus = "pci"
		device.vendor.id = "8086"
		device.vendor.name = "Intel Corporation"
		device.product.id = "9d71"
		device.product.name = "Sunrise Point-LP HD Audio"
		device.form_factor = "internal"
		device.string = "front:0"
		device.buffering.buffer_size = "705600"
		device.buffering.fragment_size = "352800"
		device.access_mode = "mmap+timer"
		device.profile.name = "analog-stereo"
		device.profile.description = "Analog Stereo"
		device.description = "Built-in Audio Analog Stereo"
		alsa.mixer_name = "Realtek ALC298"
		alsa.components = "HDA:10ec0298,17aa5062,00100103 HDA:8086280b,80860101,00100000"
		module-udev-detect.discovered = "1"
		device.icon_name = "audio-card-pci"
	ports:
		analog-output-speaker: Speakers (priority 10000, latency offset 0 usec, available: unknown)
			properties:
				device.icon_name = "audio-speakers"
		analog-output-headphones: Headphones (priority 9000, latency offset 0 usec, available: no)
			properties:
				device.icon_name = "audio-headphones"
	active port: <analog-output-speaker>
  * index: 3
	name: <bluez_sink.04_52_C7_0A_32_96.a2dp_sink>
	driver: <module-bluez5-device.c>
	flags: HARDWARE DECIBEL_VOLUME LATENCY FLAT_VOLUME 
	state: RUNNING
	suspend cause: (none)
	priority: 9950
	volume: front-left: 26212 /  40% / -23.88 dB,   front-right: 26212 /  40% / -23.88 dB
	        balance 0.00
	base volume: 65536 / 100% / 0.00 dB
	volume steps: 65537
	muted: no
	current latency: 59.72 ms
	max request: 3 KiB
	max rewind: 0 KiB
	monitor source: 4
	sample spec: s16le 2ch 44100Hz
	channel map: front-left,front-right
	             Stereo
	used by: 1
	linked by: 1
	fixed latency: 45.32 ms
	card: 3 <bluez_card.04_52_C7_0A_32_96>
	module: 26
	properties:
		bluetooth.protocol = "a2dp_sink"
		device.description = "Bose QuietComfort 35"
		device.string = "04:52:C7:0A:32:96"
		device.api = "bluez"
		device.class = "sound"
		device.bus = "bluetooth"
		device.form_factor = "headphone"
		bluez.path = "/org/bluez/hci0/dev_04_52_C7_0A_32_96"
		bluez.class = "0x240418"
		bluez.alias = "Bose QuietComfort 35"
		device.icon_name = "audio-headphones-bluetooth"
	ports:
		headphone-output: Headphone (priority 0, latency offset 0 usec, available: yes)
			properties:
				
	active port: <headphone-output>
`

The `*` seems to be in the right place, but there's certainly no visualisation output. Any ideas?

**Comment by karlstav** [comment]
Cava utilizes the "monitor" source associated with each sink. I think maybe the "monitor" source 
 is not working with all types off output. I know it's also buggy with HDMI. Try to file a bug with pulseaudio.
